### ISSUE TITLE
Skeptical about new rule prefer-toHaveBeenCalledWith

### ISSUE THREAD

**Issue by DamienCassou** [tlvince/eslint-plugin-jasmine#140]
I'm skeptical about the new rule `prefer-toHaveBeenCalledWith`. I understand that very often you want to check arguments to a spy call. But sometimes, what you care about is just that the call happened, you don't care about what were the arguments. This is for example the case when arguments are tested differently, somewhere else.

**Comment by burabure** [comment]
I don't think being explicit about the parameters that you're looking creates a problem on any test. `toHaveBeenCalled` is very misleading and it frequently hides bugs and false positives on specs.

If you have an issue using `toHaveBeenCalledWith` on some spec, I'd be happy to take a look at the snippet =)

**Comment by DamienCassou** [comment]
I should have started saying I really like the new rule `prefer-jasmine-matcher` :-).

Regarding `prefer-toHaveBeenCalledWith`, my point is: sometimes I just don't care about what arguments are passed to the spy. I would like to be able to express this. Said differently: for some tests, if I use `prefer-toHaveBeenCalledWith` and if, later, the system changes in a way the argument changes, I don't want the test to fail.

One solution could be something like:

```js
expect(spy).toHaveBeenCalledWith(jasmine.anyArguments)
```

If you insist on at least specifying the number of arguments, I would like to write:

```js
expect(spy).toHaveBeenCalledWith(jasmine.any, jasmine.any, jasmine.any)
```

Is one of those possible already? Do you know any workaround?

**Comment by DamienCassou** [comment]
Apparently, this idea

```js
expect(spy).toHaveBeenCalledWith(jasmine.any, jasmine.any, jasmine.any)
```

is somewhat already possible:

```js
expect(spy).toHaveBeenCalledWith(jasmine.anything(), jasmine.anything(), jasmine.anything())
```

but this only work for non-null and non-undefined parameters.

**Comment by burabure** [comment]
> for some tests, if I use prefer-toHaveBeenCalledWith and if, later, the system changes in a way the argument changes, I don't want the test to fail.

That is one of the issues of using `toHaveBeenCalled`, behavior can drastically change on the tested logic but tests would still pass.

If your issue is one of granularity and testing strategy (I don't care what's fired on some specs, becase I'll test it on a dedicated spec), using `jasmine.any(SomeType)` is the way to go.

**Comment by DianaSuvorova** [comment]
There is always an option to update the repo's `.eslintrc.json` file to overwrite default rule configuration. 
What I saw in other eslint plugins - The default settings tend to be permissive rather than strict. So if there is any controversial rule it would be `off` by default. 
@burabure, I think we should follow this practice and accept @DamienCassou PR. Let me know if you have any objections.

-Diana

**Comment by burabure** [comment]
@DianaSuvorova I agree that plugins should keep defaults permissive, but I don't think there's an issue on this case. What @DamienCassou is explaining seems more like a problem with a particular testing style to me than the rule being too strict. 

My reasoning is basically this: `toHaveBeenCalled` is considered a code smell because of it's vagueness and it's a proved fact that it causes false positives on specs; the plugin's job is to advise against using it.

This plugin has 7K downloads a day so if more people have issues with the rule I'd be happy to reconsider on my side.

Having said that id like to clear up that this is only my opinion, and I don't have neither the last word or exclusive control on what goes in or out of this codebase =)

**Comment by DianaSuvorova** [comment]
I don't have a strong opinion either way. Let's keep the issue open to see if there is enough signal to make the change.

**Comment by seb0zz** [comment]
@burabure do you have an example of `toHaveBeenCalled` producing false positives?

**Comment by burabure** [comment]
@seb0zz not at hand, we solved those a long time ago in my company. but it's easy to imagine how a function is called with the wrong callback or with completely wrong arguments that causes bugs and the spec still passes

**Comment by seb0zz** [comment]
Yeah I get your point but what if a method doesn't expect any parameters and you only want it to be called? Could be something like an `increment()` method? Then `toHaveBeenCalled` would make sense I think. 

Also - if `toHaveBeenCalled` would be a bad practice wouldn't it be a better approach to create an issue on Jasmine trying to get rid of the method by deprecating it?

**Comment by burabure** [comment]
>Yeah I get your point but what if a method doesn't expect any parameters and you only want it to be called? Could be something like an increment() method? Then toHaveBeenCalled would make sense I think.

like I said before, that is dependent of your code style (functional vs imperative in this example), and you can still do the same and be more specific by using `toHaveBeenCalledWith()`

> Also - if toHaveBeenCalled would be a bad practice wouldn't it be a better approach to create an issue on Jasmine trying to get rid of the method by deprecating it?

that discussion is completely out of the scope of this library, eslint rules are meant to enforce good practices, so if all libraries (and the language) did that by default, basically eslint wouldn't exist.

**Comment by DamienCassou** [comment]
Here is an example of what I mean:

```js
it('binds closeContextMenu() to the click event', () => {
  spyOn(menuManagement, 'closeContextMenu')

  let result = menuManagement.createBackButtonHeader()
  result.click()

  expect(menuManagement.closeContextMenu).toHaveBeenCalled()
})
```

in this example, I don't care that the DOM passes a mouse event to `closeContextMenu()` (this function does not even declare an `event` argument). I just want to check that the function is called.

**Comment by burabure** [comment]
@DamienCassou I'd use `jasmine.any(Event)`, but you can also use `jasmine.anything()` or if you dislike doing this and you find yourself needing it for a spec, just disable the rule for that file with a comment at the top; at last if your codebase doesn't really benefit or just don't like the rule you can just disable it on your `.eslintrc`

**Comment by DamienCassou** [comment]
@burabure: Thank you for your answer. I know about the workarounds. My point was that this new rule tags reasonable code as invalid and this annoys me. I think I explained my case enough and give a reasonable example of code. If you still disagree, you may want to close the ticket.

**Comment by burabure** [comment]
just for the sake of clearness before closing this I'll reiterate why I think this rule is important so that people searching for this have an easy time finding them before opening a new issue.
- `toHaveBeenCalled` is considered a code smell because of it's vagueness
- It causes false positives on specs
- Behavior can drastically change on the tested logic but tests would still pass
- It can (and has in my experience) hide bugs
- If you still want to be vague you can use `jasmine.any()` or `jasmine.anything()`
- This is a eslint plugin to help with best practices, our job is to warn against these things, you can override the config if you don't agree with our defaults. 

If you're reading this and you think you have strong evidence that goes against these points, please open a new issue

**Comment by jakehockey10** [comment]
I wasn't sure if this was worth opening a new issue or not, especially since this isn't a bug report.  But I'm wondering what is the intended process of handling expecting a method/function to be have been called, adhering to the prefer-toHaveBeenCalledWith rule, when the method/function is parameter-less.  I can't seem to adhere to the rule and have the test pass.  Do I just mark that line with ignore the rule this one time?  Or does jasmine have a matcher for (nothing)?

**Comment by jakehockey10** [comment]
oops.  Well I could just try not passing anything into `toHaveBeenCalledWith`.  That seems to work.  False alarm
